### PR TITLE
Protect admin endpoints against brute force attacks

### LINKED ISSUE #245: Protection against brute force attacks
Along with #244, it would be nice to have some throttling mechanism as a protection against brute force attacks, at least for the admin authentication

### PR DESCRIPTION
**Pull request by 0livd**
Add a throttling mechanism to prevent a client brute
forcing the authentication form, based on its ip address
Closes #245 

When using multiple workers, the remaining number of attempts can get inconsistent but will still be limited to num_workers * max_attempts. The solution would be to use redis to share the state between the workers but I guess we'd rather keep it simple...

### CODE REVIEW

File: ihatemoney/translations/fr/LC_MESSAGES/messages.po [hunk 1]
```diff
@@ -163,13 +163,25 @@ msgid "Export file format"
...
 msgid "This private code is not the right one"
 msgstr "Le code que vous avez entré n'est pas correct"
 
+#: web.py:106
+msgid "This admin password is not the right one. Only 2 attempts left."
+msgstr "Le mot de passe administrateur que vous avez entré n'est pas correct. Plus que 2 tentatives."
+
+#: web.py:106
+msgid "This admin password is not the right one. Only 1 attempts left."
+msgstr "Le mot de passe administrateur que vous avez entré n'est pas correct. Plus q'une seule tentative."
```
Comment (almet) [inline ihatemoney/translations/fr/LC_MESSAGES/messages.po]: typo: «plus qu'une »

File: ihatemoney/translations/fr/LC_MESSAGES/messages.po [hunk 2]
```diff
@@ -163,13 +163,25 @@ msgid "Export file format"
...
-#: web.py:95
 msgid "This private code is not the right one"
 msgstr "Le code que vous avez entré n'est pas correct"
 
+#: web.py:106
+msgid "This admin password is not the right one. Only 2 attempts left."
+msgstr "Le mot de passe administrateur que vous avez entré n'est pas correct. Plus que 2 tentatives."
+
+#: web.py:106
+msgid "This admin password is not the right one. Only 1 attempts left."
```
Comment (almet) [inline ihatemoney/translations/fr/LC_MESSAGES/messages.po]: nit: is it possible to localize the beginning of the sentence just once?

File: ihatemoney/utils.py [hunk 3]
```diff
@@ -131,3 +132,38 @@ def list_of_dicts2csv(dict_to_convert):
 
 # base64 encoding that works with both py2 and py3 and yield no warning
 base64_encode = base64.encodestring if six.PY2 else base64.encodebytes
+
+
+class LoginThrottler():
+    """Simple login throttler used to limit authentification attempts based on client's ip address
```
Comment (almet) [inline ihatemoney/utils.py]: It's « Authentication » not « Authentification ».

File: ihatemoney/utils.py [hunk 4]
```diff
@@ -131,3 +132,38 @@ def list_of_dicts2csv(dict_to_convert):
...
+    When using multiple workers, remaining number of attempts can get inconsistent
+    but will still be limited to num_workers * max_attempts
```
Comment (almet) [inline ihatemoney/utils.py]: nit: finish your sentences with a dot.

File: ihatemoney/utils.py [hunk 5]
```diff
@@ -131,3 +132,38 @@ def list_of_dicts2csv(dict_to_convert):
...
+    """
+    def __init__(self, max_attempts=3, delay=1):
+        self._max_attempts = max_attempts
+        # Delay in minutes before attempts counter is reset
```
Comment (almet) [inline ihatemoney/utils.py]: nit: «before resetting the attempts counter »

File: ihatemoney/utils.py [hunk 6]
```diff
@@ -131,3 +132,38 @@ def list_of_dicts2csv(dict_to_convert):
...
+    def increment_attempts_counter(self, ip):
+        if self._attempts.get(ip) is None:
+            # Store first attempt date and number of attempts since
+            self._attempts[ip] = [datetime.now(), 0]
+        self._attempts.get(ip)[1] += 1
+
+    def is_login_allowed(self, ip):
+        if self._attempts.get(ip) is None:
+            return True
+        # Reset counter when the delay since first attempt is over
```
Comment (almet) [inline ihatemoney/utils.py]: We can rephrase this: « When the delay is expired, reset the counter. »

File: ihatemoney/utils.py [hunk 7]
```diff
@@ -131,3 +132,38 @@ def list_of_dicts2csv(dict_to_convert):
...
+        self._delay = delay
+        self._attempts = {}
+
+    def get_remaining_attempts(self, ip):
+        return self._max_attempts - self._attempts.get(ip, [datetime.now(), 0])[1]
+
+    def increment_attempts_counter(self, ip):
+        if self._attempts.get(ip) is None:
+            # Store first attempt date and number of attempts since
+            self._attempts[ip] = [datetime.now(), 0]
```
Comment (almet) [inline ihatemoney/utils.py]: I'm a bit concerned with this behavior. I might be too worried for nothing, but this makes it possible to overwhelm the memory of the server by faking a lot of requests from different IP addresses. In order to fix this, we could free the memory (reset the dict) once we hit a big number of IP addresses (let's say 10k IPs ?). What do you think?

File: ihatemoney/web.py [hunk 8]
```diff
@@ -89,14 +91,24 @@ def admin():
...
+                # Valid password
+                if (check_password_hash(current_app.config['ADMIN_PASSWORD'],
+                                        form.admin_password.data)):
+                    session['is_admin'] = True
+                    session.update()
+                    login_throttler.reset(client_ip)
+                    return redirect(goto)
+                # Invalid password
+                login_throttler.increment_attempts_counter(client_ip)
+                msg = _("This admin password is not the right one. Only %d attempts left." %
```
Comment (almet) [inline ihatemoney/web.py]: [gettext](https://pythonhosted.org/Flask-Babel/#flask.ext.babel.gettext) can be passed the second argument directly to ease the work of translators.

**Comment by almet** [comment]
Also this might be conflicting with #247, can you build your changes on top of this branch there? Thanks!

File: ihatemoney/utils.py (hunk 7 above)
Comment (0livd) [inline ihatemoney/utils.py]: To fake random IP addresses, a client would manipulate the X-FORWARDED-FOR header. ProxyFix offers a protection against that because by default, if there are multiple proxies, it only trusts the headers set by the last one. In non-proxy setups, it means that the headers set by the client would be trusted and that the OOM killer might pay us a visit.
TLDR: Yeah let's add an auto-reset :)

**Comment by 0livd** [comment]
Changes pushed !

**Comment by 0livd** [comment]
>Also this might be conflicting with #247, can you build your changes on top of this branch there? Thanks!
Why not simply validate and merge this and then rebase #247 ?

**Comment by almet** [comment]
Simply because I already took the time to rebase it once and I don't really want to do it again :) If you feel like doing it, please go ahead, that will be appreciated!

**Comment by 0livd** [comment]
Ok I'll do it but I won't have time in the coming weeks

**Comment by almet** [comment]
Thanks a lot, and happy holidays (?) ;)

**Comment by 0livd** [comment]
>Thanks a lot, and happy holidays (?) ;)
Yep, flying to kirghistan, hope they have some decent vodka :)

**Comment by 0livd** [comment]
Back from the other side of the planet...Should we merge this? I'll then take care of the grunt work on #247.

**Comment by almet** [comment]
I won't have time to look at this in the upcoming weeks. I'll be back at the end of August!

**Comment by almet** [comment]
Yes, this looks good ! Lets merge :-)
